libc/string_cpy.c

```diff
@@ -0,0 +1,1382 @@
+		 input[MAX_STR_LEN] = TEST_STR2;
+
+	/* Trying copy zero bytes */
+	TEST_ASSERT_EQUAL_PTR(testArray, memcpy(testArray, input, 0));
+	TEST_ASSERT_EQUAL_STRING(testArray, TEST_STR1);
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_memcpy, big)
+{
+	/* Pragma must be used because of warnings with forming offset*/
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Warray-bounds"
+
+	char buff[BIG_NUMB] = { 0 };
+
+	/* Checking capability of handling big blocks of data */
+	TEST_ASSERT_EQUAL_INT(0, big_string());
+	TEST_ASSERT_NOT_NULL(memcpy(buff, long_str, sizeof(long_str)));
+	TEST_ASSERT_EQUAL_STRING(buff, long_str);
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(buff, long_str, 256);
+
+	/*
+	 * Offset big chunk of data remembering we want to
+	 * copy data from point 128 and we must trim length of cpy to it size
+	 */
+	TEST_ASSERT_NOT_NULL(memcpy(&buff[128], long_str, sizeof(long_str) - 128));
+	TEST_ASSERT_NOT_EQUAL(0, strcmp(buff, long_str));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(&buff[128], long_str, 128);
+
+#pragma GCC diagnostic pop
+}
+
+
+/*
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+ */
+
+
+TEST_SETUP(string_memccpy)
+{
+}
+
+
+TEST_TEAR_DOWN(string_memccpy)
+{
+}
+
+
+TEST(string_memccpy, basic)
+{
+	/* Ifdef used because lack of function 'memccpy' */
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+
+
+	char strDest[MAX_STR_LEN] = { 0 },
+		 strPlaceholder[MAX_STR_LEN] = TEST_STR2;
+
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(strDest, TEST_STR1, 'x', sizeof(TEST_STR1)));
+	TEST_ASSERT_EQUAL_STRING(TEST_STR1, strDest);
+
+	/* Checking if we can overwrite data using memcpy*/
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(strPlaceholder, TEST_STR1, 'x', sizeof(TEST_STR1)));
+	TEST_ASSERT_EQUAL_STRING(TEST_STR1, strDest);
+	TEST_ASSERT_EQUAL_STRING(strPlaceholder, TEST_STR1);
+
+	/* Copy only half of string str2 and in addition we searching for letter 's' */
+	TEST_ASSERT_EQUAL_PTR(&strPlaceholder[8], memccpy(strPlaceholder, TEST_STR2, 's', sizeof(TEST_STR2) / 2));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(strPlaceholder, TEST_STR2, 8);
+
+#pragma GCC diagnostic pop
+#endif
+}
+
+
+TEST(string_memccpy, stop_cpy_point_char)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char bigStrDest[BIG_NUMB] = { 0 };
+	int i;
+
+	TEST_ASSERT_EQUAL_INT(0, big_string());
+	/*
+	 * Memccpy will return a byte after found endpoint
+	 * in case of find a smaller dataset it will always point to on next byte in the array
+	 * In this case we use long_str and try to find all chars between [1 and 255]
+	 * and check if the pointer that memccpy return is the same as the point where value should be
+	 */
+	for (i = 0; i < MAX_ASCII_SET; i++) {
+		TEST_ASSERT_EQUAL_PTR(&bigStrDest[i + 1], memccpy(bigStrDest, long_str, long_str[i], MAX_ASCII_SET));
+	}
+
+	/* Search range out of bounds */
```

Comment:
I think it's a check for an another test case, we could also check all the possible values, which are not present in the string - the test case could be named `stop_char_not_found`,  Above that you could test various `n` values - so you can extend the `zero` case - rename it to `various_sizes` and add the following checks:

 - `n` that equals 0
 - `n` that is lower than stop character position,
 - `n` that is greater than stop character position
 - `n` that is greater than stop character position and has the maximum possible value